src/middlewared/middlewared/plugins/kubernetes_linux/update.py

```diff
@@ -359,10 +358,19 @@ async def validate_data(self, data, schema, old_data):
     async def validate_interfaces(self, data):
         errors = []
         interfaces = await self.route_interface_choices()
-        for k in filter(
-            lambda k: data[k] and data[k] not in interfaces, ('route_v4_interface', 'route_v6_interface')
-        ):
-            errors.append((k, data[k]))
+        interface_states = {
+            i['id']: i.get('state', {}).get('link_state') == 'LINK_STATE_UP'
+            for i in await self.middleware.call('interface.query')
+        }
+        valid_choices = [i for i in filter(lambda k: interface_states.get(k), interfaces)]
+        for k in filter(lambda k: data[k], ('route_v4_interface', 'route_v6_interface')):
+            err_str = ''
+            if data[k] not in interfaces:
+                err_str = f'Please specify a valid interface (i.e {", ".join(valid_choices)!r}).'
+            elif not interface_states.get(data[k]):
+                err_str = f'Please specify a valid interface which is active (i.e {", ".join(valid_choices)!r}).'
+            if err_str:
```

Comment:
No reason to have this check at the end. Just `errors.append` in each branch inside the loop. We should also return early on the first true evaluation so it removes risk of having verrors that have basically duplicate data. Finally `data[k]` isn't used in the caller so no reason to pass it back i.e.
```
if data[k] not in interfaces:
    errors.append((k, f'Please specify a valid interface (i.e {", ".join(valid_choices)!r}).'))
    break
elif not interfaces_states.get(data[k]):
    errors.append((k, f'Please specify a valid interface which is active (i.e {", ".join(valid_choices)!r}).'))
    break

Reply:
@yocalebo that is used but in a different place and also we don't want to break, we are looping over basically attributes which we want to validate as they are not related to each other in the sense of duplication and each should be validated separately

Reply:
The caller is doing this
```
        for k, _, err_s in await self.validate_interfaces(data):
             verrors.add(f'{schema}.{k}', err_s)
```
So how is `data[k]` being used anywhere else except inside that function?

Reply:
It's being used in another file - when we start k8s basically

Reply:
so there is another caller, i understand now 😄